sentence.go

```diff
@@ -74,69 +78,125 @@ func (s BaseSentence) TalkerID() string {
+
+	// OnTagBlock is callback to handle all parsed tag blocks even for lines containing only a tag block and
+	// allows to track multiline tag group separate lines
+	//
+	// Example of group of 3:
+	// \g:1-3-1234,s:r3669961,c:1120959341*hh\
+	// \g:2-3-1234*hh\!ABVDM,1,1,1,B,.....,0*hh
+	// \g:3-3-1234*hh\$ABVSI,r3669961,1,013536.96326433,1386,-98,,*hh
+	OnTagBlock func(tagBlock TagBlock)
+}
+
+// NewSentenceParser creates new instance of SentenceParser
+func NewSentenceParser() *SentenceParser {
+	return &SentenceParser{
+		parserLock:    &sync.RWMutex{},
+		customParsers: map[string]ParserFunc{},
+	}
+}
+
+// ParseBaseSentence parses BaseSentence from input
+func (p *SentenceParser) ParseBaseSentence(raw string) (BaseSentence, error) {
 	raw = strings.TrimSpace(raw)
-	tagBlockParts := strings.SplitN(raw, `\`, 3)
 
 	var (
 		tagBlock TagBlock
 		err      error
 	)
-	if len(tagBlockParts) == 3 {
-		tags := tagBlockParts[1]
-		raw = tagBlockParts[2]
-		tagBlock, err = parseTagBlock(tags)
+	if raw[0] == TagBlockSep {
+		// tag block is always at the start of line. Starts with `\` and ends with `\` and has valid sentence
+		// following or <CR><LF>
+		//
+		// Note: tag block group can span multiple lines but we only parse ones that have sentence
+		endOfTagBlock := strings.LastIndexByte(raw, TagBlockSep)
+		if endOfTagBlock <= 0 {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence tag block is missing '\\' at the end")
+		}
+		tagBlock, err = parseTagBlock(raw[1:endOfTagBlock])
 		if err != nil {
 			return BaseSentence{}, err
 		}
+		if p.OnTagBlock != nil {
+			p.OnTagBlock(tagBlock)
+		}
+		raw = raw[endOfTagBlock+1:]
 	}
 
 	startIndex := strings.IndexAny(raw, SentenceStart+SentenceStartEncapsulated)
 	if startIndex != 0 {
 		return BaseSentence{}, fmt.Errorf("nmea: sentence does not start with a '$' or '!'")
 	}
-	sumSepIndex := strings.Index(raw, ChecksumSep)
-	if sumSepIndex == -1 {
-		return BaseSentence{}, fmt.Errorf("nmea: sentence does not contain checksum separator")
+	checksumSepIndex := strings.Index(raw, ChecksumSep)
+	fieldsRaw := raw[startIndex+1:]
+	checksumRaw := ""
+	if checksumSepIndex != -1 {
+		fieldsRaw = raw[startIndex+1 : checksumSepIndex]
+		checksumRaw = strings.ToUpper(raw[checksumSepIndex+1:])
 	}
-	var (
-		fieldsRaw   = raw[startIndex+1 : sumSepIndex]
-		fields      = strings.Split(fieldsRaw, FieldSep)
-		checksumRaw = strings.ToUpper(raw[sumSepIndex+1:])
-		checksum    = Checksum(fieldsRaw)
-	)
-	// Validate the checksum
-	if checksum != checksumRaw {
-		return BaseSentence{}, fmt.Errorf(
-			"nmea: sentence checksum mismatch [%s != %s]", checksum, checksumRaw)
+	if p.CheckCRC == nil {
+		if checksumSepIndex == -1 {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence does not contain checksum separator")
+		}
+		if checksum := Checksum(fieldsRaw); checksum != checksumRaw {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence checksum mismatch [%s != %s]", checksum, checksumRaw)
+		}
 	}
```

Comment:
We should have a top level `func CheckCRC` with the same signature as `SentenceParser.CheckCRC`. Then this code can be written as:

``` go
checkCRC := p.CheckCRC
if checkCRC == nil {
    checkCRC = CheckCRC
}
if err := checkCRC(...); err != nil {
    return BaseSentence{}, err
}
```

Reply:
I have locally default function
```go
func DefaultCheckCRC(fieldsRaw string, sentence BaseSentence) error {
	if sentence.Checksum == "" {
		return fmt.Errorf("nmea: sentence does not contain checksum separator")
	}
	if checksum := Checksum(fieldsRaw); checksum != sentence.Checksum {
		return fmt.Errorf("nmea: sentence checksum mismatch [%s != %s]", checksum, sentence.Checksum)
	}
	return nil
}
```